proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1012,6 +1012,270 @@ lemma refillUpdate_corres:
+  done
+
+crunches non_overlapping_merge_refills
+  for valid_objs[wp]: valid_objs
+
+lemma nonOverLappingMergeRefills_valid_refills'[wp]:
+  "nonOverlappingMergeRefills scPtr \<lbrace>valid_refills' scPtr\<rbrace>"
+  apply (wpsimp simp: nonOverlappingMergeRefills_def)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+definition head_insufficient_loop_measure where
+  "head_insufficient_loop_measure sc_ptr
+     \<equiv> measure (\<lambda>(_, s). case kheap s sc_ptr of Some (Structures_A.SchedContext sc _)
+                                             \<Rightarrow> (length (sc_refills sc)))"
+
+lemma non_overlapping_merge_refills_terminates:
+  "\<lbrakk>pred_map (\<lambda>cfg. refills_unat_sum (scrc_refills cfg) \<le> unat max_time)
+             (sc_refill_cfgs_of s) sc_ptr;
+    pred_map (\<lambda>cfg. unat MIN_BUDGET \<le> refills_unat_sum (scrc_refills cfg))
+             (sc_refill_cfgs_of s) sc_ptr\<rbrakk>
+   \<Longrightarrow> whileLoop_terminates (\<lambda>_ s. the (head_insufficient sc_ptr s))
+                            (\<lambda>_. non_overlapping_merge_refills sc_ptr) r s"
+  (is "\<lbrakk>?P s; ?Q s\<rbrakk> \<Longrightarrow> _")
+  apply (rule_tac I="\<lambda>_. ?P and ?Q"
+               in whileLoop_terminates_inv[where R="head_insufficient_loop_measure sc_ptr"])
+    apply simp
+   apply (intro hoare_vcg_conj_lift_pre_fix)
+    apply (wpsimp wp: non_overlapping_merge_refills_refills_unat_sum_lower_bound
+                      non_overlapping_merge_refills_refills_unat_sum)
+    apply (fastforce dest: head_insufficient_length_at_least_two)
+   apply (wpsimp wp: update_sched_context_wp
+               simp: head_insufficient_loop_measure_def non_overlapping_merge_refills_def
+                     refill_pop_head_def update_refill_hd_def)
+   apply (fastforce dest: head_insufficient_length_at_least_two
+                    simp: vs_all_heap_simps obj_at_def)
+  apply (clarsimp simp: head_insufficient_loop_measure_def)
+  done
+
+lemma refills_unat_sum_MIN_BUDGET_implies_non_empty_refills:
+  "pred_map (\<lambda>cfg. unat MIN_BUDGET \<le> refills_unat_sum (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr
+   \<Longrightarrow> pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) sc_ptr"
+  apply (auto simp: vs_all_heap_simps refills_unat_sum_def MIN_BUDGET_nonzero unat_eq_zero)
+  done
+
+lemma headInsufficientLoop_corres:
+  "sc_ptr = scPtr
+   \<Longrightarrow> corres dc (pspace_aligned and pspace_distinct and sc_at sc_ptr and is_active_sc sc_ptr
+                  and valid_objs
+                  and (\<lambda>s. pred_map (\<lambda>cfg. unat MIN_BUDGET \<le> refills_unat_sum (scrc_refills cfg))
+                                    (sc_refill_cfgs_of s) sc_ptr)
+                  and (\<lambda>s. pred_map (\<lambda>cfg. refills_unat_sum (scrc_refills cfg) \<le> unat max_time)
+                                    (sc_refill_cfgs_of s) sc_ptr))
+                 (valid_objs' and valid_refills' scPtr)
```

Comment:
You can derive `valid_refills'` from `valid_objs'`. See `valid_objs'_valid_refills'` in StateRelation.thy.
You will need to cross `is_active_sc`, with a few hops, and there are lemmas for that, too.